File: insecureapplication-go/gallery-idp/db/clients.go
package db

import "database/sql"

type Client struct {
	ClientID     string
	Name         string
	ClientSecret string
	Trusted      bool
	CreatedAt    string
}

// CreateClient stores a client. redirectURI is intentionally accepted but
// discarded -- see the comment on the clients table in 001-init.sql: the
// original Mongoose Client schema has no redirectURIs field either, so a
// self-registered client's declared callback is never actually recorded or
// checked anywhere.
func (d *DB) CreateClient(clientID, name, secret string, trusted bool, redirectURI string) error {
	_, err := d.Exec(`INSERT INTO clients (client_id, name, client_secret, trusted) VALUES (?, ?, ?, ?)`,
		clientID, name, secret, boolToInt(trusted))
	return err
}

func (d *DB) GetClient(clientID string) (*Client, error) {
	row := d.QueryRow(`SELECT client_id, name, client_secret, trusted, created_at FROM clients WHERE client_id = ?`, clientID)
	return scanClient(row)
}

func (d *DB) ListClients() ([]*Client, error) {
	rows, err := d.Query(`SELECT client_id, name, client_secret, trusted, created_at FROM clients ORDER BY client_id`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []*Client
	for rows.Next() {
		c := &Client{}
		var trusted int
		if err := rows.Scan(&c.ClientID, &c.Name, &c.ClientSecret, &trusted, &c.CreatedAt); err != nil {
			return nil, err
		}
		c.Trusted = trusted != 0
		out = append(out, c)
	}
	return out, rows.Err()
}

// UpdateClientFields: no ownership/admin check performed by callers (IDOR --
// any logged-in user may edit any client). redirectURIs is accepted upstream
// but never lands in `allowed` since the column doesn't exist.
func (d *DB) UpdateClientFields(clientID string, fields map[string]string) error {
	allowed := map[string]bool{"name": true, "client_secret": true, "trusted": true}
	for col, val := range fields {
		if !allowed[col] {
			continue
		}
		if _, err := d.Exec(`UPDATE clients SET `+col+` = ? WHERE client_id = ?`, val, clientID); err != nil {
			return err
		}
	}
	return nil
}

func (d *DB) DeleteClient(clientID string) error {
	_, err := d.Exec(`DELETE FROM clients WHERE client_id = ?`, clientID)
	return err
}

// UpsertClientCredentials rewrites an existing client's id+secret in place,
// used at startup to honor CLIENT_ID/CLIENT_SECRET env overrides for the
// seeded "photoprint" client (mirrors mongo-seed/seed.sh's updateOne upsert).
func (d *DB) UpsertClientCredentials(oldClientID, newClientID, newSecret string) error {
	_, err := d.Exec(`UPDATE clients SET client_id = ?, client_secret = ? WHERE client_id = ?`, newClientID, newSecret, oldClientID)
	return err
}

func scanClient(row *sql.Row) (*Client, error) {
	c := &Client{}
	var trusted int
	err := row.Scan(&c.ClientID, &c.Name, &c.ClientSecret, &trusted, &c.CreatedAt)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	c.Trusted = trusted != 0
	return c, nil
}

func boolToInt(b bool) int {
	if b {
		return 1
	}
	return 0
}

File: insecureapplication/gallery/db/clients.js
const {getDb} = require('./db');
const {orNull} = require('./id');

function scanClient(row) {
  if (!row) return null;
  return {...row, trusted: !!row.trusted};
}

// Mongoose's Boolean SchemaType cast the string "false" (what a urlencoded
// form body gives you for trusted=false) to the boolean false; plain JS
// truthiness does not ("false" is a non-empty string), so raw `val ? 1 : 0`
// would mark a client trusted=1 on exactly the input meant to keep it
// untrusted. This reproduces Mongoose's cast rather than JS truthiness.
function toBoolInt(val) {
  if (typeof val === 'string') {
    return (val !== '' && val !== 'false' && val !== '0') ? 1 : 0;
  }
  return val ? 1 : 0;
}

// redirectURI is intentionally accepted but discarded -- the original
// Mongoose Client schema has no redirectURIs field either, so a
// self-registered client's declared callback is never actually recorded or
// checked anywhere.
function createClient({clientId, name, clientSecret, trusted}) {
  getDb().prepare(
      `INSERT INTO clients (client_id, name, client_secret, trusted) VALUES (?, ?, ?, ?)`,
  ).run(clientId, orNull(name), clientSecret, toBoolInt(trusted));
}

function getClient(clientId) {
  return scanClient(getDb().prepare(`SELECT * FROM clients WHERE client_id = ?`).get(clientId));
}

function listClients() {
  return getDb().prepare(`SELECT * FROM clients ORDER BY client_id`).all().map(scanClient);
}

// vulnerability: no ownership/admin check performed by callers (IDOR --
// any logged-in user may edit any client). redirectURIs is accepted
// upstream but never lands in ALLOWED_FIELDS since the column doesn't exist.
const ALLOWED_FIELDS = {name: true, client_secret: true, trusted: true};

function updateClientFields(clientId, fields) {
  for (const [col, val] of Object.entries(fields)) {
    if (!ALLOWED_FIELDS[col]) continue;
    const value = col === 'trusted' ? toBoolInt(val) : orNull(val);
    getDb().prepare(`UPDATE clients SET ${col} = ? WHERE client_id = ?`).run(value, clientId);
  }
}

function deleteClient(clientId) {
  getDb().prepare(`DELETE FROM clients WHERE client_id = ?`).run(clientId);
}

module.exports = {createClient, getClient, listClients, updateClientFields, deleteClient};
